### DISCUSSION TITLE
FRAME-FM December Sprint Summary

### DISCUSSION THREAD

**Discussion by agstephens** [NERC-EDS/FRAME-FM#29] [Announcements]
# Logistics

When: 04/12/2025, 0900-1600
Where: MS Teams
Who: 
- CEDA: AS, MP
- BGS: AK
- UKCEH: MH, JC, MDP
- NOC: TG, CS, AW

# Introduction

See our plan for the day: https://github.com/NERC-EDS/FRAME-FM/blob/sprint-dec-2025/README-Sprint-2025.md

**Morning session (0900-1230)**

- 0900-0920: Introduction to the sprint (this document)
- 0920-0930: Quick review of an example PyTorch-Lightning workflow (as a visual cue)
- 0930-1000: Plan sprint teams, include a lead and scribe for each:
   - Models: Spatial auto-encoders (static data) & Temporal auto-encoders (temporal and spatial)
   - Data: Building the data loaders
   - Monitoring (later): Testing and deciding on monitoring/visualisation tools
- 1030-1210: Investigations:
     - **Data Loaders**:
       - `.csv` loading with metadata - Colin
       - transformations - Matt
       - intermediary cached data - Matthew
       - Shapefile loading - Michael
       - Explore how Anemoi is using Zarr for its intermediate data 
     - **Modelling**:
       - logging intermediary classes: Adam
       - Modelling from Nature paper: Anastasia and Jeremy
       - Geospatial data sampling research : Tom
- 1210-1230: Discuss of our findings

**Afternoon session (1330-1600)**

- 1330-1345: Plan the afternoon
- 1345-1500: Further work
- 1500-1600: Summarisation
  - Record key findings
  - What are some of the big issues we found out today
  - Write down plans and next steps

# Summary of activities

AW: 
- looked at wrappers around PyTorch code
- using pytorch-lightning
- logging with mlflow
- basic wrapper that will work for training and validation loss
- sets up an expected working pattern
- explored standardising the data loading
  - for each dataset:
    - geared towards images, time series, shapefiles etc
    - e.g. Torch Dataset class for each

AK and JC:
- starting with TG's  `train_eurosat.py` example
  - plug in the AutoEncoder from the Nature paper
- finished Encoder for EuroSat training
- began building a decoder
  - thinking of normalising etc
  - currently only convolutional, EuroSat only at the moment
- TG's code, replaces ResNet with encoder part of auto-encoder
- looking to integrate with AW's approach
- See connected issue here: #33 

MH: 
- looking at loading Shapefiles into `xarray`
- from multi-polygons to xarray nd-array
- reading with `geopandas`
- various issues with need to convert grid but making good progress

TG:
- looking at strategies for temporal training
- quasi Monte Carlo approach:
  - rather than a random split - you have a mathematically accurate sample
- need to understand "blocking":
  - need to make sure you are training and validating across variables, spatial, temporal etc.
- looked at krigging and monte Carlo approaches 

CS: 
- got CSV files loading
- worked on concatenating together
- considered QC info and accessing extra metadata
- looked at QC data and issues related to different QC levels of data
  
MDP: 
- looking at transformations
- tested out XESMF for a general regridder
- issues with installation
- showed potential

MP:
- looking at a wireframe design for how data loaders and intermediate data might work
- got basic loader working for COG data
  - rioxarray to xarray

AS:
- explored Anemoi and how it is used to describe:
  - raw data
  - transformations
  - intermediate (transformed) data
- we might be able to re-use it as a third-party library
- see notes [here](https://github.com/NERC-EDS/FRAME-FM/discussions/25)
